chalk-solve/src/recursive/fulfill.rs

```diff
@@ -422,7 +422,13 @@ impl<'s, 'db, I: Interner> Fulfill<'s, 'db, I> {
 
         let interner = self.solver.program.interner();
 
-        if self.infer.is_trivial_substitution(interner, &subst) {
+        if self
+            .infer
+            .canonicalize(interner, &subst)
+            .quantified
+            .value
+            .is_identity_subst(interner)
```

Comment:
hmm what's the difference here? I think the intention was that `is_trivial_substitution` is supposed to do the same thing, maybe it needs to be fixed?

Reply:
Yeah I think it was supposed to do the same thing, but I'm not sure `is_trivial_substitution` can be made to work in this case. The clause we are trying to apply is 
```
clause=for<type> Implemented(A<^0.0>: Trait<^0.0>) :- ForAll<> { Implemented(^1.0: Constraint) })
```
so when determining if the clause can be applied we unify `$0` and `$1` (this is done in https://github.com/rust-lang/chalk/blob/e1234f618a9fbd28cd0080e25798bb21b790ca95/chalk-solve/src/recursive/fulfill.rs#L155). But that doesn't seem to actually update the `Substitution`, it just determines what new goal should be added and updates the `InferenceTable`. We don't end up solving that goal (`Implemented(^0.0: Constraint)`) so when `is_trivial_substitution` is called, neither `$0` or `$1` have been bound to anything. It's only after canonicalizing that you can detect the substitutions are non-trivial because of the unification needed to apply the clause.

Reply:
Hmm right, actually we canonicalize anyway directly after this. So could you just put the canonicalized substitution into a variable and use that instead of canonicalizing again in the else branch?

Reply:
Also then I think you could just remove `is_trivial_substitution`.

Reply:
done